examples/n-ary-app.kl

```diff
@@ -59,6 +59,30 @@
       [[list [_ args body]]
        [pure [list-syntax ['my-lam args body] stx]]])]])]
 
+(define-macros
+  ([flet*
+    (lambda (stx)
+      (syntax-case stx
+        [(list (_ fun body))
+         (syntax-case fun
+           [(list (name args def))
+            (syntax-case args
+              [(cons arg more-args)
+               (pure (list-syntax
+                      ('flet
```

Comment:
Woah! You [explained](https://github.com/gelisam/klister/pull/28#discussion_r374862904) that `define-macros` enables quotations in its body to refer to the macro's name. But here it looks like `define-macros` also enables `flet*` to refer to the names of macros which aren't even defined yet? How is that possible?

My first hypothesis was that unbound identifiers produced by a macro will simply refer to whatever identifier is in scope at the call site, whereas bound identifier won't change their meaning thanks to hygiene. However, this simple example proves that theory wrong:

    (define-macros
      ((m (lambda (stx)
            (pure 'unbound)))))
    
    (define unbound 2)
    
    (example m)  -- error: unknown <unbound>

The only difference between `unbound` and `flet` seems to be that `flet` is a macro, so my second hypothesis was that unbound identifiers produced by a macro may refer to a _macro_ which is in scope at the call site, but that hypothesis was also easy to disprove:

    (define-macros
      ((m (lambda (stx)
            (pure 'unbound)))))
    
    (define-macros
      ((unbound (lambda (stx)
                  (pure ''bound)))))
    
    (example m)  -- error: unknown <unbound>

So what's the difference between `flet*` and `m`?

Reply:
Ah, I get it! `flet*` is not referencing the `flet` below, it's referencing the `flet` builtin, and the `flet` below is shadowing the builtin, just like `#%app` and `lambda` do.

Reply:
That's right!